crates/bevy_ecs/src/query/cache.rs

```diff
@@ -0,0 +1,475 @@
+        matched_archetypes.intersect_with(&other.matched_archetypes);
+        let matched_storage_ids = if is_dense {
+            matched_tables
+                .ones()
+                .map(|id| StorageId {
+                    table_id: TableId::from_usize(id),
+                })
+                .collect()
+        } else {
+            matched_archetypes
+                .ones()
+                .map(|id| StorageId {
+                    archetype_id: ArchetypeId::new(id),
+                })
+                .collect()
+        };
+        CacheState {
+            archetype_generation: self.archetype_generation,
+            matched_tables,
+            matched_archetypes,
+            matched_storage_ids,
+            is_dense,
+        }
+    }
+}
+
+impl Debug for CacheState {
+    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
+        f.debug_struct("CacheState")
+            .field("matched_table_count", &self.matched_tables.count_ones(..))
+            .field(
+                "matched_archetype_count",
+                &self.matched_archetypes.count_ones(..),
+            )
+            .finish()
+    }
+}
+
+impl CacheState {
+    /// Returns the tables matched by this query.
+    pub fn matched_tables(&self) -> impl Iterator<Item = TableId> + '_ {
+        self.matched_tables.ones().map(TableId::from_usize)
+    }
+
+    /// Returns the archetypes matched by this query.
+    pub fn matched_archetypes(&self) -> impl Iterator<Item = ArchetypeId> + '_ {
+        self.matched_archetypes.ones().map(ArchetypeId::new)
+    }
+
+    /// Add a new archetype in the cache
+    fn cache_archetype(&mut self, archetype: &Archetype) {
+        let archetype_index = archetype.id().index();
+        if !self.matched_archetypes.contains(archetype_index) {
+            self.matched_archetypes.grow_and_insert(archetype_index);
+            if !self.is_dense {
+                self.matched_storage_ids.push(StorageId {
+                    archetype_id: archetype.id(),
+                });
+            }
+        }
+        let table_index = archetype.table_id().as_usize();
+        if !self.matched_tables.contains(table_index) {
+            self.matched_tables.grow_and_insert(table_index);
+            if self.is_dense {
+                self.matched_storage_ids.push(StorageId {
+                    table_id: archetype.table_id(),
+                });
+            }
+        }
+    }
+}
+
+/// [`QueryCache`] used if caching is disabled for a query.
+///
+/// We will not cache any matching archetypes for a query, so they will have to be recomputed
+/// from scratch every time.
+#[derive(Debug, Clone)]
+pub struct Uncached {
+    pub(super) is_dense: bool,
+}
+
+impl QueryCache for Uncached {
+    fn iteration_data<'s, 'a: 's, D: QueryData, F: QueryFilter>(
+        &'a self,
+        query: &QueryState<D, F, Self>,
+        world: UnsafeWorldCell,
+    ) -> IterationData<'s> {
+        let mut storage_ids = Vec::new();
```

Comment:
we might want to leave a todo or make an issue here to provide a way of reusing the allocations for an uncached query. Does flecs avoid allocating here somehow?